### PR TITLE
Replica testing

### PR DESCRIPTION
**Pull request by marceloyuela**
Finally tested creating a read replica instance.
Fixed some small issues in documentation.
Created the XML documentation for New-GcSqlInstanceReplicaConfig

### CODE REVIEW

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs [hunk 1]
```diff
@@ -28,7 +57,7 @@ public class NewGcSqlInstanceReplicaConfigCmdlet : GcSqlCmdlet
         /// </para>
         /// </summary>
         [Parameter]
-        public SwitchParameter FailoverTarget { get; set; }
+        public bool FailoverTarget { get; set; } = false;
```
Comment (ILMTitan) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Why did you make this  a false defaulted bool?

Comment (marceloyuela) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Woops, this was something I did when debugging. Will fix

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs [hunk 2]
```diff
@@ -114,7 +115,7 @@ public class NewGcSqlInstanceReplicaConfigCmdlet : GcSqlCmdlet
         /// </para>
         /// </summary>
         [Parameter]
-        public SwitchParameter MySqlVerifyCertificate { get; set; }
+        public bool MySqlVerifyCertificate { get; set; } = false;
```
Comment (marceloyuela) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Fragment from debugging. Will go back to switchparameter.

**Comment by ILMTitan** [comment]
I am a little confused about why you changed FailoverTarget from a switch parameter to a bool that defaults to false.

**Comment by marceloyuela** [comment]
Fixed that fragment from debugging.

File: Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1 [hunk 3]
```diff
@@ -73,6 +73,24 @@ Describe "Add-GcSqlInstance" {
         $myInstance.Settings.MaintenanceWindow.Hour | Should Be 2
         gcloud sql instances delete $instance --quiet 2>$null
     }
+
+    It "should be able to create a read replica instance" {
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1]: Should this be written as "read-replica"?

File: Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1 [hunk 4]
```diff
@@ -73,6 +73,24 @@ Describe "Add-GcSqlInstance" {
...
+        $replicaConfig = New-GcSqlInstanceReplicaConfig
+        $config = New-GcSqlInstanceConfig $instance -SettingConfig $setting `
+            -ReplicaConfig $replicaConfig -MasterInstanceName "test-db2"
+        Add-GcSqlInstance $config
+        $newInstances = Get-GcSqlInstance
+        ($newInstances.Name -contains $instance) | Should Be true
+        $myInstance = Get-GcSqlInstance $instance
+        $myInstance.MasterInstanceName | Should Be "gcloud-powershell-testing:test-db2"
+        $myInstance.InstanceType | Should Be "READ_REPLICA_INSTANCE"
+        gcloud sql instances delete $instance --quiet 2>$null
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1]: Have we implemented the corresponding cmdlet by now? Perhaps we should just use that?

File: Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1 [hunk 5]
```diff
@@ -73,6 +73,24 @@ Describe "Add-GcSqlInstance" {
...
+        $r = Get-Random
+        # A random number is used to avoid collisions with the speed of creating
+        # and deleting instances.
+        $instance = "test-inst$r"
+        $setting = New-GcSqlSettingConfig "db-n1-standard-1" 
+        $replicaConfig = New-GcSqlInstanceReplicaConfig
+        $config = New-GcSqlInstanceConfig $instance -SettingConfig $setting `
+            -ReplicaConfig $replicaConfig -MasterInstanceName "test-db2"
+        Add-GcSqlInstance $config
+        $newInstances = Get-GcSqlInstance
```
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1]: Maybe add a line of vertical whitespace here for readability? Currently this is a big wall of text.

File: Google.PowerShell/Sql/GcSqlInstanceConfigCmdlets.cs [hunk 6]
```diff
@@ -149,6 +150,13 @@ protected override void ProcessRecord()
                     Name = FailoverReplica
                 };
             }
+            if (MasterInstanceName != null)
+            {
+                //This should only be specified for read replica instances, so we make some adjustments.
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlInstanceConfigCmdlets.cs]: Space after `//`.

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs [hunk 7]
```diff
@@ -13,6 +13,35 @@ namespace Google.PowerShell.Sql
     /// Creates a configuration for a replicaConfiguration. 
     /// Can be pipelined into New-GcSqlInstanceConfig.
     /// </para>
+    /// <para>
+    /// WARNING: If a parameter passed in is bad, or wrong, this cmdlet will not error.
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Having the WARNING is great. But perhaps we can reword this as "a parameter passed in is bad, or wrong" could just be "any parameter is incorrect.

Can we validate these parameters somehow? Is there a command we can issue that doesn't modify the project but will get the errors? If not, no worries.

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs [hunk 8]
```diff
@@ -13,6 +13,35 @@ namespace Google.PowerShell.Sql
...
+    /// <example>
+    ///   <para>
+    ///   Creates a basic replica configuration resource.
+    ///   </para>
+    ///   <para><code>
+    ///     PS C:\> New-GcSqlInstanceReplicaConfig
+    ///   </code></para>
+    ///   <br></br>
+    ///   <para>
+    ///   (If successful, the command returns a ReplicaConfiguration resource containing the replica configuration.)
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Why not put this above on line 23?

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs [hunk 9]
```diff
@@ -78,10 +107,11 @@ public class NewGcSqlInstanceReplicaConfigCmdlet : GcSqlCmdlet
         /// <summary>
         /// <para type="description">
         /// Interval in milliseconds between replication heartbeats.
+        /// Defaults to 100.
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: I assume we specify the default here because the parameter is required in the underlying API?

100ms seems far too small. There is a performance cost of the overhead of a heartbeat. I'd put this to at least 5s. Maybe as high as 20 or 30s.

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs [hunk 10]
```diff
@@ -133,7 +163,8 @@ protected override void ProcessRecord()
                     Password = MySqlPassword,
                     SslCipher = MySqlSslCipher,
                     Username = MySqlUser,
-                    VerifyServerCertificate = MySqlVerifyCertificate
+                    VerifyServerCertificate = MySqlVerifyCertificate,
+                    Kind = "sql#mysqlReplicaConfiguration"
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Why is this necessary? Seems like an API wart. We shouldn't have to deal with this.

File: Google.PowerShell/Sql/GcSqlSettingConfigCmdlet.cs [hunk 11]
```diff
@@ -76,7 +76,7 @@ public enum ActivationPolicy
         /// <para type="description">
         /// Whether binary log is enabled.
         /// If backup configuration is disabled, binary log must be disabled as well.
-        /// Defaults to true;
+        /// Defaults to true for non replica instances.
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlSettingConfigCmdlet.cs]: non-replica instances.

File: Google.PowerShell/Sql/GcSqlSslCmdlets.cs [hunk 12]
```diff
@@ -479,6 +480,10 @@ protected override void ProcessRecord()
                 default:
                     throw UnknownParameterSetException;
             }
+            if (!ShouldProcess($"{project}/{instance}", "Reset SSL Certificate Configuration"))
+            {
+                return;
```
Comment (chrsmith) [inline Google.PowerShell/Sql/GcSqlSslCmdlets.cs]: Using  `if (!X) { return; }` should only be reserved if the following code block is very long. Since it is only 4-lines, I'd just put it in an `if (ShouldProcess(...)) {` block.

**Comment by chrsmith** [comment]
Some more feedback.

File: Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1 (hunk 4 above)
Comment (marceloyuela) [inline Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1]: Won't that couple the testing of this cmdlet even more with a different cmdlet?  Is that okay?

File: Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs (hunk 8 above)
Comment (marceloyuela) [inline Google.PowerShell/Sql/GcSqlInstanceReplicaConfigCmdlet.cs]: Will do, and will do for the rest of the XML docs as well

File: Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1 (hunk 4 above)
Comment (chrsmith) [inline Google.PowerShell.IntegrationTests/SQL/SQL.GcSqlInstances.Tests.ps1]: It is possible that you could have a cycle where cmdlet A calls cmdlet B calls cmdlet A, and if there is a bug in either of them then the test could be considered invalid.

However, I assume that `Remove-GcSqlInstance` has a separate set of tests that work great. And so it might make sense to just use that rather than rely on `gcloud`.

**Comment by marceloyuela** [comment]
Should have fixed the above issues

**Comment by marceloyuela** [comment]
Any other comments?

**Comment by chrsmith** [comment]
LGTM if @ILMTitan is good with it.

**Comment by ILMTitan** [comment]
LGTM
